Create FamilyTree.java, Human.java, Main.java

IPaulPI opened the pull request.

IPaulPI committed:
Create FamilyTree.java, Human.java, Main.java

Liberate520 commented on src/family_tree/FamilyTree.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    // Метод для отображения всего генеалогического дерева
+    public void displayFamilyTree() {
```
правильней будет возвращать текст и для этого есть метод toString

Liberate520 commented on src/family_tree/Human.java:
```diff
@@ -0,0 +1,39 @@
+public class Human {
+    private String name; // Имя человека
+    private String birthDate; // Дата рождения
+    private List<Human> children; // Список детей этого человека
```
также рекомендуется добавить список родителей

Liberate520 commented on src/family_tree/Human.java:
```diff
@@ -0,0 +1,39 @@
+// Класс, описывающий человека в генеалогическом дереве
+public class Human {
+    private String name; // Имя человека
+    private String birthDate; // Дата рождения
```
еще дату смерти для расчета возраста

Liberate520 replied:
и пол человека

Liberate520 commented on src/family_tree/Human.java:
```diff
@@ -0,0 +1,39 @@
+    // Метод для добавления ребенка к этому человеку
+    public void addChild(Human child) {
+        this.children.add(child);
+    }
```
добавить toString

IPaulPI committed:
Create FileHandler.java, Serializable.java and I made changes to the rest of the classes, taking into account the comments

Liberate520 commented on src/family_tree/FileHandler.java:
```diff
@@ -0,0 +1,58 @@
+
+import java.io.*;
+
+public class FileHandler {
```
ожидается наличие интерфейса по заданию

Liberate520 commented on src/family_tree/Serializable.java:
```diff
@@ -0,0 +1,4 @@
+package family_tree;
+
+public interface Serializable {
```
Создание этого интерфейса ошибка. Интерфейс с таким названием уже есть и наличие дубля может привести в проблемам

IPaulPI committed:
The Iterable interface for the tree is implemented.
Methods have been created to sort the list of people before output
A package structure has been created for the project

IPaulPI committed:
The MyIterable interface for the tree is implemented.
Methods have been created to sort the list of people before output
A package structure has been created for the project

IPaulPI committed:
The Serializable interface has been removed.

IPaulPI committed:
The Serializable interface has been removed.

IPaulPI committed:
The Serializable interface has been removed.

IPaulPI committed:
Merge remote-tracking branch 'origin/master'

Liberate520 commented on src/ru/gb/family_tree/FamilyTree.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.*;
+
+// Класс, описывающий генеалогическое дерево
+public class FamilyTree implements Serializable, MyIterable<Human> {
```
К сожалению интерфейс MyIterable ничего не даст вам. Вам нужно реализовать интерфейс Iterable из пакета java для возможности использования объекта в цикле for-each.

IPaulPI replied:
Спасибо!

IPaulPI committed:
The Serializable interface has been removed.

IPaulPI committed:
Generics have been added to the FamilyTree class.

IPaulPI committed:
Generics have been added to the FamilyTree class.

Liberate520 commented on src/ru/gb/family_tree/FamilyTree.java:
```diff
@@ -0,0 +1,42 @@
+        return new ArrayList<>(members.values());
+    }
+
+    public List<T> sortBy(Comparator<T> comparator) {
```
метод сортировки не должен ожидать компаратора в качестве аргумента. Дерево должно предоставлять компараторы в методах сортировки, а не ожидать их извне

IPaulPI committed:
Implemented MVP pattern.

IPaulPI committed:
Implemented MVP pattern.

IPaulPI committed:
The project with a family tree was refactored taking into account SOLID principles.

Liberate520 commented on src/ru/gb/family_tree/presenter/FamilyTreePresenter.java:
```diff
@@ -0,0 +1,49 @@
+import java.io.Serializable;
+
+public class FamilyTreePresenter<T extends HasRelations<T> & Serializable> {
+    private FamilyTree<T> familyTree;
```
нужно создать Сервисный класс для объединения работы модели

Liberate520 commented on src/ru/gb/family_tree/presenter/FamilyTreePresenter.java:
```diff
@@ -0,0 +1,49 @@
+public class FamilyTreePresenter<T extends HasRelations<T> & Serializable> {
+    private FamilyTree<T> familyTree;
+    private FamilyTreeView view;
+
```
добавьте методы сохранения и загрузки

Liberate520 commented on src/ru/gb/family_tree/view/ConsoleView.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+import java.util.Scanner;
+
+public class ConsoleView<T> implements FamilyTreeView<T> {
```
view в mvp отвечает не только за вывод данных, но и за ввод данных. Класс должен обеспечивать двусторонний диалог с пользователем
